protocol/message.go

```diff
@@ -104,3 +107,114 @@ func NewMonitoringProof(ap []*m.AuthenticationPath,
 	str []*m.SignedTreeRoot, e ErrorCode) (*DirectoryProofs, ErrorCode) {
 	return &DirectoryProofs{MonitoringType, ap, str, e}, e
 }
+
+// Verify implementation
+
+func (res *ErrorResponse) Verify(string, []byte, sign.PublicKey) ErrorCode {
+	return res.Error
+}
+
+func verifyAuthPath(uname string,
+	ap *m.AuthenticationPath,
+	str *m.SignedTreeRoot) ErrorCode {
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return ErrorBadVRFProof
+	}
+
+	// verify commitment
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) && /* proof of inclusion */
+		!ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+		return ErrorBadCommitment
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return ErrorBadMapping
+	}
+
+	return Passed
+}
+
+func verifySTR(signKey sign.PublicKey,
+	str *m.SignedTreeRoot,
+	savedSTR []byte) ErrorCode {
+
+	// verify STR's signature
+	if !signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	if str.Epoch > 0 && savedSTR != nil &&
+		!m.VerifyHashChain(str.PreviousSTRHash, savedSTR) {
+		return ErrorBadSTR
+	}
+
+	return Passed
+}
+
+func (df *DirectoryProof) Verify(uname string, savedSTR []byte, signKey sign.PublicKey) ErrorCode {
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	if e := verifySTR(signKey, str, savedSTR); e != Passed {
+		return e
+	}
+
+	// verify TB's Signature
+	if tb != nil {
+		if !signKey.Verify(tb.Serialize(str.Signature), tb.Signature) {
+			return ErrorBadSignature
+		}
+	}
+
+	if e := verifyAuthPath(uname, ap, str); e != Passed {
+		return e
+	}
+
+	// verify TB's VRF index
+	if tb != nil {
+		if !bytes.Equal(tb.Index, ap.LookupIndex) {
+			return ErrorBadIndex
+		}
+	}
+
+	return Passed
+}
+
+func (df *DirectoryProofs) Verify(uname string, savedSTR []byte, signKey sign.PublicKey) ErrorCode {
+	if len(df.AP) < 1 ||
+		(len(df.AP) > 1 && len(df.AP) != len(df.STR)) {
+		return ErrorMalformedDirectoryMessage
+	}
+
+	for i := 0; i < len(df.STR); i++ {
```

Comment:
`KeyLookupInEpoch` returns `DirectoryProofs`, too, right? But this set of STR checks wouldn't work for a `KeyLookupInEpoch` response as we've discussed in https://github.com/coniks-sys/coniks-go/issues/70#issuecomment-243852256, right?